beater/java_attacher/java_attacher.go

```diff
@@ -22,61 +22,389 @@ import (
+
+	// trying to improve start time accuracy - worth to consider optimizing as this runs every time the loop runs (1 second or so)
+	if j.executeForEachJvm(ctx, jvms, j.obtainAccurateStartTime, false, time.Second) {
+		j.logger.Errorf("timeout trying to find accurate start time for %v jvms", len(jvms))
+	}
+
+	j.filterCached(jvms)
+
+	if j.executeForEachJvm(ctx, jvms, j.obtainCommandLineArgs, false, time.Second) {
+		j.logger.Errorf("timeout trying to find command line args for %v jvms", len(jvms))
+	}
+
+	// todo: decide whether we have enough advantage to do that within the integration instead of doing it within the attacher
+	j.filterByDiscoveryRules(jvms)
+
+	if j.executeForEachJvm(ctx, jvms, j.verifyJvmExecutable, true, time.Second) {
+		j.logger.Errorf("timeout trying to verify Java executables for %v jvms", len(jvms))
+	}
+
+	for _, jvm := range jvms {
+		j.logger.Debugf("found a Java process (PID %v) of version %v that was started at %v by '%v' using command '%v'",
+			jvm.pid, jvm.version, jvm.startTime, jvm.user, jvm.command)
 	}
-	return JavaAttacher{
-		cfg:    cfg,
-		logger: logger,
-	}, nil
+	return jvms, nil
 }
 
-// javaAttacher is bundled by the server
-var javaAttacher = filepath.FromSlash("./java-attacher.jar")
+func (j *JavaAttacher) discoverAllRunningJavaProcesses(ctx context.Context) (map[string]*JvmDetails, error) {
+	jvms := make(map[string]*JvmDetails)
+
+	// We use `start`, which is very inaccurate only as a fallback, expecting it to be replaced with an accurate start time later on.
+	// An example for the `start` time format: "4:01PM". We may decide that with the combination of PID, it is safe enough for
+	// uniqueness determination for caching purposes, especially if we remove stale PIDs when we don't see them.
+	// We avoid using `lstart` here because it is not one of the standard ps columns, and because it may have different format in different
+	// systems. In addition, it would complicate output parsing as its format is full of spaces, which we rely on as a delimiter between
+	// output line parts.
+	cmd := exec.CommandContext(ctx, "ps", "-A", "-ww", "-o", "user,uid,gid,pid,start,comm")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	scanner := bufio.NewScanner(stdout)
+	for scanner.Scan() {
+		line := scanner.Text()
+		psOutputLineParts := strings.Fields(line)
+		if len(psOutputLineParts) > 5 {
+			command := psOutputLineParts[5]
+			if strings.Contains(strings.ToLower(command), "java") {
+				pid := psOutputLineParts[3]
+				jvms[pid] = &JvmDetails{
+					psOutputLineParts[0],
+					psOutputLineParts[1],
+					psOutputLineParts[2],
+					pid,
+					psOutputLineParts[4],
+					psOutputLineParts[5],
+					"unknown",
+					"",
+				}
+			}
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		j.logger.Errorf("error reading output for command '%v': %v", strings.Join(cmd.Args, " "), err)
+	}
+	err = cmd.Wait()
+	return jvms, err
```

Comment:
Very useful feedback, thanks!

Reply:
Since there are other errors that I want to return **instead of a list**, I will keep the signature, but make sure I either return non-nil error or non-nil list. I believe that follows the convention.